app/src/main/kotlin/compiler/parser/analysis/GrammarAnalysis.kt

```diff
@@ -1,10 +1,41 @@
 package compiler.parser.analysis
 
+import compiler.common.dfa.state_dfa.Dfa
+import compiler.common.dfa.state_dfa.DfaState
 import compiler.parser.grammar.AutomatonGrammar
+import compiler.parser.grammar.Production
 
 class GrammarAnalysis<S : Comparable<S>> {
     fun computeNullable(grammar: AutomatonGrammar<S>): Set<S> {
-        return TODO()
+        val nullable: MutableSet<S> = HashSet()
+        val stateQueue: ArrayDeque<DfaState<S, Production<S>>> = ArrayDeque()
+        val conditionalSets: MutableMap<S, MutableSet<DfaState<S, Production<S>>>> = HashMap()
+
+        grammar.productions.forEach { (_, dfa) -> stateQueue.addAll(dfa.getAcceptingStates()) }
+        fun analyzePredecessors(state: DfaState<S, Production<S>>, dfa: Dfa<S, Production<S>>) {
+            val predecessorsMap = dfa.getPredecessors(state)
+            for ((edge, predecessors) in predecessorsMap) {
+                if (edge in nullable) {
+                    stateQueue.addAll(predecessors)
```

Comment:
Can't this bfs loop forever? I think if we found some nullable symbol quickly and then there was some cycle in the dfa containing only these nullable symbols on edges, we'd keep spinning infinitely over that cycle, adding predecessors in `analyzePredecessors`

Reply:
Good catch, I'll fix that!